### PR TITLE
Refresh docs for GitHub Actions usage.

### PR DESCRIPTION
**Pull request by ScottTodd**
Progress on https://github.com/iree-org/iree/issues/16203 and https://github.com/iree-org/iree/issues/17957.

New page preview: https://scotttodd.github.io/iree/developers/general/github-actions/

## Overview

This documents the state of our GitHub Actions usage after recent refactoring work. We used to have a monolithic `ci.yml` that ran 10+ builds on every commit, many of which used large CPU runners to build the full project (including the compiler) from source. Now we have platform builds, some of which run every commit but many of which run on nightly schedules. Most interesting _test_ jobs are now a part of "pkgci", which builds release packages and then runs tests using those packages (and sometimes runtime source builds).

Net results:

* Infrastructure costs should be more manageable - far fewer workflow runs will require powerful CPU builders
* Self-hosted runners with special hardware have dedicated places to be slotted in and are explicitly not on the critical path for pull requests

## Details

* Add new README badges
  * pre-commit, OpenSSF Best Practices, releases (stable, nightly), PyPI packages, select CI workflows
* Add github-actions.md webpage
  * Full workflow status tables
  * Descriptions of each type of workflow
  * Tips on (not) using Docker
  * Information about `pull_request`, `push`, `schedule` triggers
  * Information about GitHub-hosted runners and self-hosted runners
  * Maintenance/debugging tips
* Small cleanup passes through related files

skip-ci: modified script does not affect builds

### CODE REVIEW

File: docs/website/docs/developers/general/github-actions.md [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+    * Jobs most representative of core developer workflows should aim to run
+      here. Jobs can be marked [required](#required-and-optional-checks) or
+      [opt-in on presubmit](#opt-in-for-presubmit-jobs) on a case-by-case basis.
+* `push`
+    * Jobs running here should be a superset of jobs running on `pull_request`.
+* `schedule`
+    * Jobs designed to run nightly (e.g. nightly releases), jobs for non-core
+      configurations (like certain sanitizers/fuzzers), and jobs using
+      self-hosted runners in low supply can run on a schedule instead of on
+      every commit.
```
Comment (ScottTodd) [inline docs/website/docs/developers/general/github-actions.md:201]: I could add a few examples here explaining the current state.

* Why ci_linux_x64_clang_asan.yml is running on `pull_request` but ci_linux_x64_clang_tsan.yml is running on `schedule` (asan is extremely useful, tsan is only situationally useful, both require big CPU runners)
* Why ci_windows_x64_msvc.yml is running on `schedule` (standard GitHub-hosted runners are slow, large GitHub-hosted runners are expensive, our CPU builder pool only has Linux runners in it)
* Why ci_linux_arm64_clang.yml is running on `schedule` (Needs either large GitHub-hosted runners or self-hosted, and we had availability issues with the self-hosted runners)

File: docs/website/docs/developers/general/github-actions.md [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+      reliably, and how access is limited to the group that maintains the
+      hardware, any jobs using these self-hosted runners must be optional and
+      easy to disable.
+
+!!! info - "Contributing self-hosted runners"
+
+    Want to run tests on your own hardware as part of IREE's upstream CI? Get
+    in touch with us on one of our
+    [communication channels](../../index.md#communication-channels) and we'd be
+    happy to help discuss the options available.
```
Comment (ScottTodd) [inline docs/website/docs/developers/general/github-actions.md]: Could give some rough statistics here for how many runners, which how many cores, etc. would make sense for the various event triggers / support tiers (see the recent discussion at https://groups.google.com/g/iree-discuss/c/8gcs4otsrVw/)

File: docs/website/docs/developers/general/github-actions.md [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+* Testing packages across a variety of platforms.
+* Publishing the <https://iree.dev> website.
+* Updating dependencies.
+
+Workflows are defined directly in the repository at
+[`.github/workflows/`](https://github.com/iree-org/iree/tree/main/.github/workflows),
+using a mix of official actions
+(e.g. [actions/checkout](https://github.com/actions/checkout)),
+community actions
+(e.g. [pre-commit/action](https://github.com/pre-commit/action)),
```
Comment (marbre) [inline docs/website/docs/developers/general/github-actions.md]: Are the pre-commit actions referred to as community actions? Googling for `github community actions` get's me to the marketplace which is aligned with my expectations.

File: docs/website/docs/developers/general/github-actions.md [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+* _Workflows_ are configurable automated processes that run one or more _jobs_.
+* _Jobs_ are a set of _steps_ in a workflow that are executed on the same
+  _runner_.
+* _Steps_ are lists of commands or meta actions to run in a shell environment.
+* _Runners_ are servers (physical or virtual machines) that run _workflows_
+  when triggered.
+* _Events_ are specific activities in a repository that trigger a
+  _workflow run_.
+
+<img src="https://docs.github.com/assets/cb-25535/mw-1440/images/help/actions/overview-actions-simple.webp">
```
Comment (marbre) [inline docs/website/docs/developers/general/github-actions.md]: I dislike to point this out but we're probably shouldn't embedded content that is served on a site with `© 2024 GitHub, Inc.` and  which is not under a CC-BY or whatever license, especially as we have a `Copyright © 2024 The IREE Authors` on our website. However, should be straight forward to recreate.

File: docs/website/docs/developers/general/github-actions.md [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+
+GitHub supports
+[`paths` and `paths-ignore` filters](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onpushpull_requestpull_request_targetpathspaths-ignore)
+for `push` and `pull_request` events that can be used to configure which
+workflows run based on paths modified. This mechanism is simple but somewhat
+limited in what it can express, so we have a custom mechanism for marking
+certain jobs as conditionally enabled:
+
+* All jobs run on `push` events, after pull requests are merged ('postsubmit').
+* Jobs may be marked as opt-in for `pull_request` events ('presubmit') by editing
```
Comment (marbre) [inline docs/website/docs/developers/general/github-actions.md]: Should 'postsubmit' and 'presubmit'  be highlighted more clearly? Or should apostrophes be removed?

File: docs/website/docs/developers/general/github-actions.md [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+      [autoscaling](https://docs.github.com/en/actions/hosting-your-own-runners/managing-self-hosted-runners/autoscaling-with-self-hosted-runners)),
+      or persistent. Persistent runners can retain local build and artifact
+      caches to improve workflow time substantially.
+
+!!! info - "Contributing self-hosted runners"
+
+    Want to run tests on your own hardware as part of IREE's upstream CI? Get
+    in touch with us on one of our
+    [communication channels](../../index.md#communication-channels) and we'd be
+    happy to help discuss the options available.
```
Comment (marbre) [inline docs/website/docs/developers/general/github-actions.md]:
```suggestion
    happy to discuss the options available.
```
?

File: docs/website/docs/developers/general/github-actions.md [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+
+## :material-book-open-outline: Maintenance tips
+
+* Certain workflow failures are posted in the `#github-ci` channel in IREE's
+  Discord server
```
Comment (marbre) [inline docs/website/docs/developers/general/github-actions.md]:
```suggestion
  Discord server.
```
I think most listings are ended with a trailing punctuation so far.

File: docs/website/docs/developers/general/github-actions.md [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+* GitHub supports
+  [notifications for workflow runs](https://docs.github.com/en/actions/monitoring-and-troubleshooting-workflows/monitoring-workflows/notifications-for-workflow-runs)
+  but only for the user that last modified a workflow
```
Comment (marbre) [inline docs/website/docs/developers/general/github-actions.md]:
```suggestion
  but only for the user that last modified a workflow.
```

File: docs/website/docs/developers/general/github-actions.md [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+* GitHub sometimes experiences incidents outside of our control. Monitor
+  <https://www.githubstatus.com/> to see if issues are widespread.
+* Repository admins can monitor repository runners at
+  <https://github.com/iree-org/iree/settings/actions/runners>
```
Comment (marbre) [inline docs/website/docs/developers/general/github-actions.md]:
```suggestion
  <https://github.com/iree-org/iree/settings/actions/runners>.
```

File: docs/website/docs/developers/general/github-actions.md [hunk 10]
```diff
@@ -0,0 +1,315 @@
...
+* Organization admins can monitor organization runners at
+  <https://github.com/organizations/iree-org/settings/actions/runners>
```
Comment (marbre) [inline docs/website/docs/developers/general/github-actions.md]:
```suggestion
  <https://github.com/organizations/iree-org/settings/actions/runners>.
```

File: docs/website/docs/developers/general/github-actions.md [hunk 11]
```diff
@@ -0,0 +1,315 @@
...
+* Guides for GitHub Actions: <https://docs.github.com/en/actions/guides>
+* Events that trigger workflows: <https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows>
+* About GitHub-hosted runners: <https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners>
+* About large runners: <https://docs.github.com/en/actions/using-github-hosted-runners/about-larger-runners/about-larger-runners>
+* GitHub Actions Runner application: <https://github.com/actions/runner>
+* GitHub Actions Runner Images with included software: <https://github.com/actions/runner-images>
+
+Community:
+
+* A curated list of awesome things related to GitHub Actions: <https://github.com/sdras/awesome-actions>
```
Comment (marbre) [inline docs/website/docs/developers/general/github-actions.md]:
```suggestion
* A curated list of awesome things related to GitHub Actions: <https://github.com/sdras/awesome-actions>.
```

**Review by marbre — CHANGES_REQUESTED** [review]
Great write up @ScottTodd. I did not go through all the CI scripts behind for this review (or rather the linked issues) but from what I was already aware of before I don't see anything wrong in here. Just some minor suggestions and nits.

File: docs/website/docs/developers/general/github-actions.md (hunk 3 above)
Comment (ScottTodd) [reply docs/website/docs/developers/general/github-actions.md]: I was mainly trying to differentiate between actions hosted at https://github.com/actions that are "official" and maintained by Microsoft/GitHub and others from the community. Official actions are generally updated with security patches and features more regularly, so they should be preferred over community actions. I can reword a bit or just leave this off.

File: docs/website/docs/developers/general/github-actions.md (hunk 4 above)
Comment (ScottTodd) [reply docs/website/docs/developers/general/github-actions.md]: Fair point, yeah. I want to limit the number of asset files we keep in the repository at https://github.com/iree-org/iree/tree/main/docs/website/docs/assets to keep the repository size more manageable. I do like having an image or diagram here to explain the concepts and break apart the walls of text. Could try porting the diagram to mermaid.js or host the image ourselves.

File: docs/website/docs/developers/general/github-actions.md (hunk 7 above)
Comment (ScottTodd) [reply docs/website/docs/developers/general/github-actions.md]: Yep, will change these all, thanks.

General guidelines I've been trying to follow: https://developers.google.com/style/lists#capitalization-and-end-punctuation (but I'm usually inconsistent)

File: docs/website/docs/developers/general/github-actions.md (hunk 3 above)
Comment (marbre) [reply docs/website/docs/developers/general/github-actions.md]: I see your point more clearly now but at a first glance https://github.com/pre-commit/action looked quite "official" even though it is of course not. Maybe replace the URL and point to the marketplace (https://github.com/marketplace?type=actions)?

Comment (ScottTodd) [reply docs/website/docs/developers/general/github-actions.md]: Just removed the sentence about where we source our actions from.

File: docs/website/docs/developers/general/github-actions.md (hunk 4 above)
Comment (ScottTodd) [reply docs/website/docs/developers/general/github-actions.md]: Converted to mermaid.js
image

File: docs/website/docs/developers/general/github-actions.md (hunk 5 above)
Comment (ScottTodd) [reply docs/website/docs/developers/general/github-actions.md]: Added an info section explaining the terms then dropped the quotes.

**Comment by ScottTodd** [comment]
Made a few edits in response to review comments. PTAL.

**Review by marbre — APPROVED** [review]
Looks good to me 👍

**Pull request merged**